### ISSUE TITLE
Notebook diff not compatible with readonly files from TextDocumentContentProvider - only checks IFileService for a IFileSystemProvider

### ISSUE THREAD

**Issue by rehmsen** [microsoft/vscode#209017]
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: vscode.dev 
- OS Version: macOS Sonoma 14.4.1 (23E224)

## Steps to Reproduce:

1. Have an extension that registers a `TextDocumentContentProvider` with a different "scheme", like "my-scheme://".
2. Try to open a notebook diff where the left side is served via "my-scheme://". 

## Expected:
The notebook diff can be opened with a left hand side provided via a TextDocumentContentProvider.

## Actual:
It fails in this line, because it requires the fileService to have aIFileSystemProvider for it:
https://github.com/microsoft/vscode/blob/a1d4ad18475f82b729e92f52d22eef140cfef5d4/src/vs/workbench/contrib/notebook/browser/services/notebookServiceImpl.ts#L176

## Preliminary investigation and assumptions:

My (possibly incorrect) understanding of the provided abstractions and concepts is:
* TextDocumentContentProvider is for readonly documents (such as some historic version in a VCS): https://github.com/microsoft/vscode/blob/a1d4ad18475f82b729e92f52d22eef140cfef5d4/src/vscode-dts/vscode.d.ts#L1798-L1806
* FileSystemProvider is for file systems that allow reading and writing: https://github.com/microsoft/vscode/blob/a1d4ad18475f82b729e92f52d22eef140cfef5d4/src/vscode-dts/vscode.d.ts#L8856-L8867
* The left hand side of a diff cannot be edited - this should be true both for regular and notebook diffs.

The `TextDocumentContentProvider`s seem to be registered with the `ITextModelService` -  I think Notebook diff should possibly check that instead of the FileService to see if the "my-scheme://" is supported? This check seems to have been added here: https://github.com/microsoft/vscode/issues/122627#issuecomment-829040413
https://github.com/microsoft/vscode/pull/122658

Maybe @jrieken can advise on how to get the notebook diff viewer to work with readonly files from `TextDocumentContentProvider` schemes?

**Comment by rehmsen** [comment]
I tried to dig a bit to better understand how to make the notebook diff work with a TextDocumentContentProvider instead of the FileService, but there are a lot of abstractions I do not understand. In regular diff, non-file scheme URIs seem to be loaded using a `TextResourceEditorInput` as opposed to a `FileEditorInput`. For notebook diffs, the callgraph is roughly 
NotebookEditorInput -> NotebookModelResolverServiceImpl -> NotebookModelReferenceCollection -> SimpleNotebookEditorModel -> StoredFileWorkingCopyManager -> StroedFileWorkingCopy -> FileService

Somewhere between those one would need to add support for TextDocumentContentProvider provided text, but as I don't understand what all these abstractions are for, it is difficult for me to pinpoint where exactly. Just from the names, I would say that the left hand side of a diff is not necessarily a "working copy" (it may be read only original version), so I am guessing that SimpleNotebookEditorModel should maybe have a way to branch out to the TextDocumentContentProvider provided document, but that is a wild guess.

Can someone point me in the right direction? I am happy to make a contribution, but might need some help figuring out how these things are meant to work together.

**Comment by a-stewart** [comment]
Hi,

I've been investigating this a bit and the interesting part is the `FileWorkingCopyManager`. As best as I could tell it is something that is only in use by Notebooks and provides an alternate approach for how to read files.

There is some context in https://github.com/microsoft/vscode/issues/117873 and the initial implementation is in https://github.com/microsoft/vscode/pull/119233.

As far as I understand, the desire was to have a basic underlying implementation to abstract file reading so that multiple different types of editor wouldn't need to deal with the low level logic. However, at the moment only Notebooks make use of this.

The problem is with the `TextDocumentContentProvider` API. The `FileWorkingCopyManager` handles "stored files" and "untitles files", with stored files being read write files on disk, and with untitled files being scratch buffers and files that are yet to be saved.

As far as I can see there are two things that need to happen in order for files provided by `TextDocumentContentProvider` to be supported in notebooks.

1. `FileWorkingCopyManager` needs to be aware of registered `TextDocumentContentProvider`s.
2. `FileWorkingCopyManager` needs to have a third type of file that it can consider, which are readonly files with no "stored" backing.

The approach I am looking into for the second part is to to just create a sibling of `IStoredFileWorkingCopyManager` and `IUntitledFileWorkingCopyManager` to handle files provided by `TextDocumentContentProvider`s. (Would appreciate suggestions with regards to what this should be called.)

For the first part there are two options, one would be to load in the `ITextModelService` and use the file contents are provided by that class. This feels a little bit odd to be wiring it into here, but this is currently the source of truth for documents provided by `TextDocumentContentProvider`.

Another option, would be to duplicate the calls to register a `TextDocumentContentProvider`, and to register with both `ITextModelService` (used by TextEditors) and `FileWorkingCopyManager` (used by Notebooks), but having 2 sources of truth for this feels wrong.

A third option would be to extract the handling of `TextDocumentContentProvider` into another class, and have that wiring into both the classes. This would probably be the cleanest approach, but would be a fairly large refactor, and would make the PR larger than either of the other two options.

I'm happy to work on trying to resolve this setup, and would be interested to know what your preferred approach would be, and if you would be willing to accept a PR for this.

**Comment by a-stewart** [comment]
I've had a bit of a play with this and I have a working proof of concept which adds support for the TextDocumentConentProvider API in Notebooks: https://github.com/microsoft/vscode/pull/212697.

I'd be interested in discussing this, and seeing what you think of this approach so that we can work towards creating a PR to resolve this.

**Comment by a-stewart** [comment]
@jrieken - quick ping on this.

We'd like to work on creating a solution for this, and were wondering if you could take a look at the approach in the POC in https://github.com/microsoft/vscode/pull/212697 and let us know if that seems like the best approach.

Happy to look into other approaches if you think that another way would be better.

**Comment by jrieken** [comment]
Sorry, this isn't really my domain but from what I know it is wrong to use the "TextDocument"ContentProvider for anything notebook related. The latter works with [serializers](https://github.com/microsoft/vscode/blob/a5e41db5c77b1f6cad85c049a4cf8a5927be8a71/src/vscode-dts/vscode.d.ts#L14956) and isn't text-based. This makes the marriage of both worlds sound wrong. I understand the appeal in using the simple text document content provider for this but having a simple, readonly-only file system provider is also fairly simple and that would power notebooks the way it is intened

**Comment by a-stewart** [comment]
This makes sense from a holistic view point, but the assumption is that the extension using TextDocumentContentProvider is the same extension as the one handling the notebooks.

In our case, the TextDocumentContentProvider is a source control extension, providing a view of the file at an older revision, which is used for creating diffs of a file.

We also have a notebook extension which is completely unrelated and developed by a different team.

It seems to be mixing these concerns? Why should the Notebook extension stop working based on how a completely unrelated source control extension offers file contents?

Or why should the source control extension have to go for a more complex approach to serving files on the chance that a user may also have an extension installed that uses notebooks.

And if we have this limitation with the TextDocumentContentProvider - should we document this telling authors of source control extensions or other extensions that using it will prevent the content from rendering in a notebook?

**Comment by jrieken** [comment]
> Or why should the source control extension have to go for a more complex approach to serving files on the chance that a user may also have an extension installed that uses notebooks.

Yes, it should do that in order to support other, non text document file types like notebooks or images


> And if we have this limitation with the TextDocumentContentProvider - should we document this telling authors of source control extensions or other extensions that using it will prevent the content from rendering in a notebook?

Yeah, it seems we should improve the doc. I was under the assumption that its name and its signature makes this clear but I open for suggestions on how to underline this stronger

**Comment by a-stewart** [comment]
> > Or why should the source control extension have to go for a more complex approach to serving files on the chance that a user may also have an extension installed that uses notebooks.
> 
> Yes, it should do that in order to support other, non text document file types like notebooks or images

Images are an interesting thing I hadn't considered, although VS Code doesn't currently support diffing images, which means as far as I am aware, for all current purposes, a TextDocumentContentProvider does work for a filesystem backing diffs in all ways other than diffs.

In general, I think my opinion is that TextDocumentContentProvider can provide text content for a virtual document. And Notebooks can often handle text content. In this case where text is both provided and accepted, I think that the least surprising things for VS Code would be to pass one through into the other?

> > And if we have this limitation with the TextDocumentContentProvider - should we document this telling authors of source control extensions or other extensions that using it will prevent the content from rendering in a notebook?
>
> Yeah, it seems we should improve the doc. I was under the assumption that its name and its signature makes this clear but I open for suggestions on how to underline this stronger

I think part of the issue is that it is listed very prominently under the Virtual Documents section of the extension guide, with the full filesystem only being a section at the end.

https://code.visualstudio.com/api/extension-guides/virtual-documents

The way the guide is written makes it sound like TextDocumentContentProvider is the correct approach for rendering arbitrary text files given a url with no conceptual filesystem behind it and no way to query things like file metadata or directory contents.

As for the name making it clear, I think the bit that surprised me was to hear that Notebooks aren't backed by a Text Document. I always assumed that most notebooks would be backed by json, so anything that can provide content to a text document, could also provide for a notebook too.

---

In general I would agree with everything you said from a overall point of view, if it were a single product controlling the notebook renderers and the filesystem providers. But given we have completely separate entities creating the filesystem providers and creating notebook renderers - and from both of their points of view they can handle/create the content being requested, it feels like it is VS Code in the middle that is failing to link them?

**Comment by jrieken** [comment]
> Images are an interesting thing I hadn't considered, although VS Code doesn't currently support diffing images, 

It does by showing a split editor with the original and modified state side by side. That's actually how it works for all custom editors. 


>  they can handle/create the content being requested, it feels like it is VS Code in the middle that is failing to link them?

This link is a file system provider. I guess we should make it more clear in the SCM documentation that original states should be served via FS because only that will support the full breath of features

**Comment by a-stewart** [comment]
Fair enough - I wonder if the vscode api could provide a helper method, where you just provide a provider for the file contents given a uri, and it constructs a readonly filesystem around that?

Then yeah, I think that just documenting the TextDocumentContentProvider and perhaps the Source Control APIs could be enough to steer extension authors into the right direction?

**Comment by jrieken** [comment]
>  I wonder if the vscode api could provide a helper method, where you just provide a provider for the file contents given a uri, and it constructs a readonly filesystem around that?

We don't do helpers and utils in the API, but anyone is free to do so as a npm module

**Comment by a-stewart** [comment]
I have been looking into using a readonly filesystem; the two methods I am concerned with are `stat` and `readDirectory`.

For readDirectory, since we aren't every rendering a tree with it, as far as I can see there is no issue in just returning an empty array - from experimenting with using it to back diffs it seems to never be called.

For stat though, this seems to be called before we call readFile every time, so I was wondering if it was safe to avoid calling readFile twice, and just do something like:

```ts
  stat(): vscode.FileStat {
    return {
      type: vscode.FileType.File,
      ctime: 0,
      mtime: new Date().getTime(),
      size: 0
    };
  }
```

Do we use the size reported by stat for anything important? It seems to be working as expected in diffs but I am unsure if there might be some surprising results? Do we use mtime or ctime to assist with caching?
